Major Tracing Refactor Part Deux, this time it's for keeps

kbwestfall opened the pull request:
Please have a look back at PR #741.  Most of the details are the same.

The main additions here are merging some changes to PCA coefficient fitting and trace fitting/following that Joe introduced.  Changes to the trace fitting/following are implemented explicitly in `core.trace.fit_trace` (what was `core.extract.iter_tracefit`).  However, in this PR I've removed all weighting for the PCA coefficient fits for slit tracing.  The object tracing still uses Joe's edited code with his changes to how the coefficients are weighted there.  The major next step is to fully integrate what is now basically duplicated code.

Note that there are a ton of doc files that changed.  Sorry for the noise.  Some of it is a result of me cleaning up the docs directory, and some of it is because `sphinx-apidoc` changed its spacing (I'm using sphinx==2.1.2). 

Tests pass (except for cooked).  Dev suite currently running.

kbwestfall commented:
Tests pass.  The three failures in `test_scripts.py` are because my ginga installation is fubar'd.  Cooked (v0.11.1) is uploaded to the team drive, @profxj is going to check the `test_script.py` failures.
```
test_arc.py .                                           [  0%]
test_biasframe.py ....                                  [  2%]
test_bpmimage.py ..                                     [  4%]
test_calibrationimage.py .                              [  4%]
test_calibrations.py ...........                        [ 11%]
test_coadd.py .                                         [ 11%]
test_cooked.py .                                        [ 12%]
test_flatfield.py .                                     [ 12%]
test_flexure.py .                                       [ 13%]
test_flux.py ....                                       [ 15%]
test_fluxspec.py ...                                    [ 17%]
test_load.py ..                                         [ 18%]
test_load_images.py ...........                         [ 25%]
test_masterframe.py .                                   [ 25%]
test_match.py ....                                      [ 28%]
test_metadata.py .....                                  [ 31%]
test_moment.py ......                                   [ 34%]
test_msgs.py ..                                         [ 35%]
test_opticalmodel.py ....                               [ 38%]
test_parse.py ...                                       [ 40%]
test_processimage.py ..                                 [ 41%]
test_procimg.py .                                       [ 41%]
test_pydl.py .                                          [ 42%]
test_pypeit.py ..                                       [ 43%]
test_pypeitimage.py ....                                [ 45%]
test_pypeitpar.py ......................                [ 58%]
test_pypeitsetup.py .......                             [ 62%]
test_qa.py .                                            [ 63%]
test_save.py ..                                         [ 64%]
test_scienceimage.py ....                               [ 67%]
test_scripts.py FF.F..                                  [ 70%]
test_setups.py .............                            [ 78%]
test_slitmask.py .                                      [ 78%]
test_specobj.py ..                                      [ 80%]
test_specobjs.py ....                                   [ 82%]
test_spectrographs.py ............                      [ 89%]
test_traceimage.py ..                                   [ 90%]
test_traceslits.py .....                                [ 93%]
test_utils.py ...                                       [ 95%]
test_vcorr.py ..                                        [ 96%]
test_wavecalib.py .                                     [ 97%]
test_waveimage.py .                                     [ 97%]
test_wavemodel.py .                                     [ 98%]
test_wavetilts.py ...                                   [100%]
```

kbwestfall commented:
Dev Suite passes:

--- PYPEIT DEVELOPMENT SUITE PASSED 30/30 TESTS ---

profxj commented:
pypeit/tests/test_scripts.py ......                                           [ 70%]

scripts passed on my machine with Ginga

profxj commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+            msgs.error('File does not exit: {0}'.format(filename))
+        msgs.info('Loading EdgeTraceSet data from: {0}'.format(filename))
+        with fits.open(filename) as hdu:
+            this = cls(load_spectrograph(hdu[0].header['PYP_SPEC']),
```
You might use my somewhat hacky `items_from_master_file()` method
in masterframe.py to "standardize" things.  But I'll forgive you if you don't.

kbwestfall replied:
I need to pass the whole hdu to `_reinit`, so I'd rather keep it as is, instead of changing `items_from_master_file`.  But we should probably come back to this.  Added a comment.

profxj commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+                    continue
+                _spec = spec[:,i][indx]
+                _cen = cen[:,i][indx]
+                plt.text(_spec[_spec.size//2], _cen[_spec.size//2], str(self.traceid[i]),
```
Should we axe this?  The PNG is rarely useful compared
to what ginga provides.  And if ginga is broken (ahem, Kyle)
we can/should provide a matplotlib option.

kbwestfall replied:
The difference is that ginga always requires a hands-on interaction, whereas these PNGs can be produced and stored.  I put `qa_plot` function in early on in the refactor, and then basically always preferred the `show` method because it was faster and I didn't need a hands-off qa method.  But for large runs of the pipeline on lots of data, we may eventually want these quick spot check plots that can be brought up just by looking at a webpage.  Then people can use the better ginga interface for detailed checks of something that looks weird.  These plots are a lot different than the plots that were created previously because they provide zoomed in views of each trace.

kbwestfall replied:
Bottom line is that I don't think it hurts to leave it in for now.  We can remove it later if it's not useful.

kbwestfall replied:
I've changed the code so that if `qa_path` is not `None` in the `EdgeTraceSet` object, the call to `auto_trace` will try to execute `qa_plot` and write the plots to the qa path.

profxj commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+                    continue
+                _spec = spec[:,i][indx]
+                _cen = cen[:,i][indx]
+                plt.text(_spec[_spec.size//2], _cen[_spec.size//2], str(self.traceid[i]),
```
Can you explain why you tend to unpack these as the start of these methods?
I rather like seeing them with `par` attached so that I know they can be set.

kbwestfall replied:
Yeah, this was just an artifact of the development process.  I starting by having everything be arguments, and then decided what arguments should be in the parameter list.  Instead of rewriting the code then to use the `par` stuff explicitly, I just pulled what I needed from the parameters at the beginning.  Generally, I'd rather we be using the `par` directly instead of passing off to a middle man.  I'm actually a bit torn though between that and liking the clarity  of collecting what parameters are used at the beginning of each method...

kbwestfall replied:
I understand the complaint about the variable and parameter names being different.  It was more clear to me to have the name of the variable match the keyword argument of the relevant function instead of the parameter name.  Another useful aspect of collecting all the variables extracted from `par` is you can see the variable to parameter mapping.  I'd rather make the effort to use `par` directly, but I'm leaving this for now.

profxj commented on pypeit/scripts/chk_edges.py:
```diff
@@ -17,26 +17,28 @@ def parser(options=None):
 
+    if not pargs.old:
+        edges = edgetrace.EdgeTraceSet.from_file(pargs.root)
+        edges.show(thin=10, in_ginga=True)
```
Maybe make the ginga bit optional so that people without a functional
ginga can still use this.

profxj commented on pypeit/spectrographs/gemini_gmos.py:
```diff
@@ -91,9 +91,10 @@ def default_pypeit_par():
         """
         par = pypeitpar.PypeItPar()
-        # Set wave tilts order
         par['calibrations']['slits']['sigdetect'] = 20.
```
shall we remove all the 'slits' par entries?
seems like it..

kbwestfall replied:
Sorry to continue to punt on these, but this was another todo for the clean-up PR.

profxj commented on pypeit/spectrographs/gemini_gmos.py:
```diff
@@ -133,6 +134,35 @@ def default_pypeit_par():
+
+        headarr = self.get_headarr(scifile)
+
+        # Turn PCA off for long slits
```
Note this hack of mine for longslit.  I think it is ok..

kbwestfall replied:
I worry a bit about this, but as long as 'arcsec' always picks out the longslit data, I think this is fine...

profxj commented on pypeit/tests/test_moment.py:
```diff
@@ -0,0 +1,157 @@
+import pytest
```
New tests!  I love it.

kbwestfall replied:
Thanks!

profxj commented on pypeit/tests/test_traceslits.py:
```diff
@@ -8,135 +8,102 @@
 import glob
```
maybe rename this file test_edgetrace.py ?

kbwestfall replied:
done

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -19,6 +19,7 @@
 from pypeit import flatfield
 from pypeit import traceimage
 from pypeit import traceslits
```
Just commented for now.

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -572,12 +573,14 @@ def get_flats(self):
-                    except:
-                        self.traceSlits.save(traceImage=self.mstrace)
+#                    # Add tweaked boundaries to the MasterTrace file
+#                    self.traceSlits.tslits_dict = self.flatField.tslits_dict
```
I'm going to put this off until the full integration PR.

kbwestfall commented on pypeit/core/moment.py:
```diff
@@ -0,0 +1,520 @@
+    # Window for the integration for each coordinate
+    i1 = np.floor(_col - _radius + 0.5).astype(int)
+    i2 = np.floor(_col + _radius + 0.5).astype(int)
+    c = i1[:,None]-1+np.arange(int(np.amax(np.amin(i2-i1)-1,0))+4)[None,:]
```
You're consistent :)  You had the same comment before.  The `4` isn't strictly necessary; it has to be `2`.  The `4` just ensures there are values with a weight of `0` at either end of the integration window.  This was inconsistent between the `trace_gweight` and `trace_fweight` code.  We could set to `2`, but it's a marginal different in performance.  Added a comment about this.

kbwestfall commented on pypeit/core/skysub.py:
```diff
@@ -1053,6 +1055,10 @@ def ech_local_skysub_extract(sciimg, sciivar, mask, tilts, waveimg, global_sky,
 
         # update the FWHM fitting vector for the brighest object
         indx = (sobjs.ech_objid == uni_objid[ibright]) & (sobjs.ech_orderindx == iord)
+#        try:  # MagE debuggin
```
Done.

kbwestfall commented on pypeit/core/trace.py:
```diff
@@ -0,0 +1,1499 @@
+    for ii in range(median_iterations):
+        sqmstrace = ndimage.median_filter(sqmstrace, size=(7, 3))
+
+    # Make sure there are no spuriously low pixels
```
Done.

kbwestfall commented on pypeit/core/trace.py:
```diff
@@ -0,0 +1,1499 @@
+
+    Args:
+        edge_img (`numpy.ndarray`_):
+            An array marked with -1 for left slit edges and +1 for right
```
Yeah, I think this was wrong, and I confused myself trying to understand why I would have written it this way.  So I just now give the expected shape.  I've tried to clean up the description in a few places in this file.

kbwestfall commented on pypeit/core/trace.py:
```diff
@@ -0,0 +1,1499 @@
+                    len(_cen[best]),l))
+
+        # Set the reference spatial locations to use for tracing the
+       
```
Done.  Also I added a comment to the docs that basically shouldn't be used and one should instead raise the `peak_thresh` value.  Not sure this `peak_clip` option is useful in the end.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+    """
+    Mask bits used during slit tracing.
+    """
+    # TODO: Create a script that will dynamically write the used bits
```
On the TODO list.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+        - centroid_refine
+        - fit_refine (calls fit_trace)
+        - build_pca
+        - peak_refine (calls peak_trace, which uses both pca_trace and fit_trace)
```
Thanks for pointing this out.  I had forgotten that the top level doc for EdgeTraceSet was a mess.  I've cleaned this up and documented the auto-tracing algorithm.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+            A list of strings indicating the main methods applied
+            when tracing.
+    """
+#    master_type = 'Trace'   # For MasterFrame base
```
I'm leaving this kind of stuff in for now.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+
+        Args:
+            rows (:obj:`int`, optional):
+                Number of rows for each column. If None, the table
```
For `empty_design_table`, this is the number of slits.  For `empty_objects_table`, it's the number of objects.  I've clarified this in the docstrings.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+                                 description='Slit width (arcsec)'),
+                    table.Column(name='SLITPA', dtype=float, length=length,
+                                 description='Slit position angle onsky (deg from N through E)'),
+                    table.Column(name='ALIGN', dtype=np.int16, length=length,
```
Yep; now specified in the column description.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+                                      max_ocol=self.nspat-1, extract_width=extract_width,
+                                      mask_threshold=mask_threshold)
+
+    def auto_trace(self, img, bpm=None, det=1, binning=None, save=False, debug=False,
```
I've changed the code a bit to make sure that `det` is not `None`, but I've left it as having a default of 1.  I prefer the idea of people reducing data with a single detector as not having to always specify `det=1`.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+        elif self.par['sync_predict'] == 'pca':
+            # TODO: This causes the code to fault. Maybe there's a way
+            # to catch this earlier on?
+            msgs.error('Sync predict cannot use PCA because too few edges were found.  Either choose '+msgs.newline()+
```
Added to the error message.  I originally it just change to `nearest` and go on, but this puts the pypeit file and the parameters actually used by the code out of sync.  This can cause the code to barf when you read in a saved EdgeTraceSet master file because I explicitly check that the parameters used to construct the file are the same as expected given the parameters passed in the instantiation of the object.  This is me being strict (and doesn't come up if you use the `from_file` method), but I think this is probably a case where we want to be strict.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+            - Follows the detected left and right edges along
+              spectrally adjacent pixels to identify coherent traces.
+              See :func:`pypeit.core.trace.identify_traces`.
+            - Performs basic handling of orphaned left or right
```
It isn't at the moment, but I kept it around for testing because I wasn't sure that there was some corner case that I was messing up by skipping these steps.  If you're confident we don't need that functionality, I'm fine to remove it.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4122 @@
+            self.binning = binning
+
+        # TODO: keep the TraceImage object instead of deconstructing
+        # it?  For direct input, use a base PypeItImage object
```
I like the idea of collecting a lot of the arguments that would be passed into first making a `PypeItImage`, but we could continue to allow the more base arguments but collect the relevant attributes into a `self.pypeitimage` that's kept and instantiated internally.  I don't have a strong preference either way.

jhennawi commented on pypeit/scripts/trace_edges.py:
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
```
Thanks this script is great!

kbwestfall replied:
PS -- But I've never actually tested it on tracing a single image...

jhennawi commented on pypeit/spectrographs/spectrograph.py:
```diff
@@ -773,6 +771,7 @@ def slit2order(self, slit_spat_pos):
             return self.orders[iorder], indx[iorder]
 
 
+    # TODO : This code needs serious work.  e.g. eliminate the try/except
```
I'm not following why this was changed, and why the documentation describing what this function does is now missing.

kbwestfall replied:
I think this is another change that slipped in from @profxj 's other branch.  Any comment on this @profxj  ?

jhennawi commented on pypeit/tests/test_scripts.py:
```diff
@@ -6,28 +6,80 @@
 matplotlib.use('agg')  # For Travis
 
-from pypeit import msgs
 from pypeit.scripts import setup, show_1dspec, coadd_1dspec, chk_edges, view_fits, chk_flats
```
Can we add a test of the pypeit_trace_edges script?

kbwestfall replied:
Yep, good suggestion.  Done.

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -639,14 +641,26 @@ def get_slits(self, redo=False, write_qa=True):
-                                                master_key=self.master_key_dict['trace'],
-                                                master_dir=self.master_dir, qa_path=self.qa_path,
-                                                reuse_masters=self.reuse_masters, msbpm=self.msbpm)
+        # TODO: Leave this for now for testing
```
It gets removed in my next PR.

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -639,14 +641,26 @@ def get_slits(self, redo=False, write_qa=True):
+            self.edges.load()
+            self.tslits_dict = self.edges.convert_to_tslits_dict()
+        else:
+
```
Next PR.

kbwestfall commented on pypeit/core/extract.py:
```diff
@@ -2561,6 +2562,12 @@ def ech_objfind(image, ivar, slitmask, slit_left, slit_righ, inmask=None, spec_m
                                                  debug=show_pca)
         # Trial and error shows weighting by S/N instead of S/N^2 performs better
 
```
This was from early testing of code used to remove duplicated code.  Addressed in the next PR.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+                                   master_dir=master_dir, img=traceImage, det=det, auto=True)
+    edges.save()
+
+For a command-line script that executes the automatic tracing, use `pypeit_trace_edges`.
```
Good suggestion.  Done.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
This specifies the string as a "raw string"; see here:

https://docs.python.org/3.7/reference/lexical_analysis.html#string-and-bytes-literals

I use this when I use the sphinx `:math:` command for latex-like equations and variables in the docstring.  I'm not sure if this is the only way to do this, but it's what works.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
Specified: "Find spectrally short traces"

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
Specified: "Minimum spectral length of any trace (pixels)"

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
Added this to the description of the parameter in the `EdgeTracePar`, which gets propagated to the documentation: "Most useful for echelle data or multi-slit data where the slits should have similar or identical lengths."


"Most useful for echelle data ' \
                                'or multi-slit data where the slits have similar lengths.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
See previous response on this.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
No, but is there any reason not to?

kbwestfall commented on pypeit/metadata.py:
```diff
@@ -1586,7 +1586,7 @@ def write_pypeit(self, ofile, ignore=None, cfg_lines=None, write_bkg_pairs=False
                       currently the only option.**
             configs (str, optional):
-                Configs to
+                Configs to (TODO: to what?)
```
Done.

kbwestfall commented on pypeit/par/util.py:
```diff
@@ -565,26 +565,8 @@ def make_pypeit_file(pypeit_file, spectrograph, data_files, cfg_lines=None, setu
-#        _cfg_lines += ['[calibrations]']
-#        _cfg_lines += ['    [[arcframe]]']
-#        _cfg_lines += ['        number = 1']
+    # TODO: Bring back checks for the appropriate number of calibration
```
These `number` keywords are never actually used, which is why we removed these lines.  The comment is more about whether or not we should bring back the check that one has a sufficient number of frames of each type.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+                different, a warning is issued and the bitmask
+                defined by the header is used instead of the existing
+                :attr:`bitmask`.
+            rebuild_pca (:obj:`bool`, optional):
```
As you might expect, I disagree.  Let's discuss.  

For the moment, I've put in some code that checks the versions of the python environment against those logged in the file header.  It currently throws warnings so the user will know to be careful or redo the reduction if the versions are different.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
Done.

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -0,0 +1,4186 @@
+            right = self.traceid > 0
+            cen = self.spat_cen if self.spat_fit is None else self.spat_fit
+            ginga.show_slits(viewer, ch, cen[:,np.invert(right)], cen[:,right],
+                  
```
Rows are the first axis, and columns are the second axis.  No ambiguity needed.  Maybe it's just me, but this is much more clear than spec,spat.  Having said that, I've gone through and changed variable names, comments, and docstrings to make the usage clear.

kbwestfall commented:
I've addressed all the comments now, with a couple of contentious topics that we can talk about on the telecon.  On Joe's specific complaints:

1. I've added some code that checks the relevant python environment versions against those logged in the header.  It can throw warnings (as it does now for `EdgeTraceSet`) or raise an exception.  We can be as strict or lax with this as we want.  The point of the `load` function in `EdgeTraceSet` is to put it back in the state it was when it was saved.  I agree about the possibility that the regenerated PCA may not be the same as when the data was saved.  I'll play around with this, but I'm not changing this now.
2. I've clarified that `row` and `col` means in `edgetrace.py` and/or changed the relevant text or object names.  `row` and `col` are clear to me.  First and second axis.
3. The `auto_trace` method now calls `qa_plot` at the end, if the `qa_path` is defined.

Tests pass, modulo the ginga errors in `test_scripts.py`:

```
test_arc.py .                                                            [  0%]
test_biasframe.py ....                                                   [  2%]
test_bpmimage.py ..                                                      [  4%]
test_calibrationimage.py .                                               [  4%]
test_calibrations.py ...........                                         [ 11%]
test_coadd.py .                                                          [ 11%]
test_cooked.py .                                                         [ 12%]
test_edgetrace.py .....                                                  [ 15%]
test_flatfield.py .                                                      [ 15%]
test_flexure.py .                                                        [ 16%]
test_flux.py ....                                                        [ 18%]
test_fluxspec.py ...                                                     [ 20%]
test_load.py ..                                                          [ 21%]
test_load_images.py ...........                                          [ 28%]
test_masterframe.py .                                                    [ 28%]
test_match.py ....                                                       [ 30%]
test_metadata.py .....                                                   [ 33%]
test_moment.py ......                                                    [ 37%]
test_msgs.py ..                                                          [ 38%]
test_opticalmodel.py ....                                                [ 40%]
test_parse.py ...                                                        [ 42%]
test_processimage.py ..                                                  [ 43%]
test_procimg.py .                                                        [ 44%]
test_pydl.py .                                                           [ 45%]
test_pypeit.py ..                                                        [ 46%]
test_pypeitimage.py ....                                                 [ 48%]
test_pypeitpar.py ......................                                 [ 61%]
test_pypeitsetup.py .......                                              [ 65%]
test_qa.py .                                                             [ 66%]
test_save.py ..                                                          [ 67%]
test_scienceimage.py ....                                                [ 69%]
test_scripts.py .FF.F..                                                  [ 73%]
test_setups.py .............                                             [ 81%]
test_slitmask.py .                                                       [ 81%]
test_specobj.py ..                                                       [ 83%]
test_specobjs.py ....                                                    [ 85%]
test_spectrographs.py ............                                       [ 92%]
test_traceimage.py ..                                                    [ 93%]
test_utils.py ...                                                        [ 95%]
test_vcorr.py ..                                                         [ 96%]
test_wavecalib.py .                                                      [ 97%]
test_waveimage.py .                                                      [ 97%]
test_wavemodel.py .                                                      [ 98%]
test_wavetilts.py ...                                                    [100%]
```

I need to rerun the dev suite and will post when that's done.

jhennawi commented:
Looks great. Will approve when the new set of tests pass, just past the result.

kbwestfall commented:
Dev Suite passes:

--- PYPEIT DEVELOPMENT SUITE PASSED 30/30 TESTS ---

profxj commented:
Merging.  Bring on the next PR
